README.md
# 请求库

一种封装了一些常见的业务逻辑的请求库，如请求参数与返回数据处理，业务逻辑码（retcode）处理，失败重试，统一成功失败处理等。默认支持小程序与 WEB 请求。

## 特征

- baseUrl 设置
- 失败后自动 2 次重启
- 业务返回码 retcode 字段指定及白名单处理
- request 请求参数插件处理
- response 返回数据插件处理
- 不论是成功还是失败都要处理，见下面的 completeHandler（如 hideLoading 处理，上报处理）
- 成功或失败的处理，见下面的 thenHandler 和 catchHandler
- 支持取消发送请求
- 请求适配器自定义
- 其他定制

## 安装

```bash
npm i g-request --save
```

## 使用

### 1、业务实例化并处理公有逻辑

```ts
// gRequest.ts
// ------------------------------------
import { Request, ICtx, RequestError, IInstanceOptions, defaultConfig } from 'g-request';

// 默认参数 defaultConfig 为以下的值：
// {
//   ext: {
//     baseUrl: '',
//     xRequestId: true, // header 头部加上 x-request-id，方便查看日志，默认开启
//     xRequestTime: true, // 记录请求耗时，默认开始
//     repeatNum: 2, // 默认失败自动重试 2 次
//     timeout: 0, // 默认请求本身支持 timeout 参数的不需要这个，设置为 0 即可，如果需要兼容不支持的，设置这个即可，不要设置请求本身的
//     retcodeKey: 'retcode', // retcode 字段，false 表示不启用该功能
//     retcodeWhiteList: [], // retcode 白名单，默认空数组为retcode为0进入then，其余为进入catch，false 表示不启用该功能，retcodeKey 为 false 也没有该功能
//     logicErrorMsgKey: 'message', // ，默认逻辑错误文本字段，支持一层对象，如 errData.msg
//     LoginErrorMsgUnknown: 'Unknown Error', // 逻辑错误文本默认，如果后台没有返回具体的错误，则显示该文本
//     adapter: getDefaultAdapter(), // 自动适配 WEB 还是小程序发送请求
//   },
// }

// 实例化
const gReqeust = new Request(options: IInstanceOptions);

// 插件是处理请求参数，可使用多个 use，每个 use 函数按注册顺序依次进入管道处理，最后返回 ctx
// 如小程序可再次设置 header 的 cookie 字段，用于鉴权
gRequest.req.use((ctx: ICtx) => {
  // 处理 req ...
  // const { req } = ctx;

  // 最后记得 return ctx
  return ctx;
});

// 插件式处理返回数据，同上请求参数的处理
gRequest.res.use((ctx: ICtx) => {
  // 处理 res ...
  // ctx.res = {};

  // 最后记得 return ctx
  return ctx;
});

// 请求完成处理，不论成功或失败，可用于关闭 loading，上报等
// err 有三种 type，分别为：逻辑错误，服务器错误，网络错误，具体见错误说明部分
gRequest.completeHandler = (ctx: ICtx, err?: RequestError) => {};

// 统一成功处理，如后台数据下划线转驼峰处理
export interface IRes = {
  retcode: number;
  data: IAnyObject;
  message: string;
}
gRequest.thenHandler = <IRes>(ctx: ICtx) => {
  // 返回下一步进入成功的数据
  const res: IRes = {...};
  return res;
};

// 统一错误处理：如 toast 提示，上报错误等
gRequest.catchHandler = (err: RequestError, ctx: ICtx) => {
  // 处理逻辑

  // 最后一定要返回错误，用于串起错误链
  return Promise.reject(err);
};

// 导出
export default gReqeust;
```

### 2、具体发送请求

导入 `gRequest.ts`，调用其方法发送具体请求

```js
// api.ts
// ------------------------------------
import { IRequestOptions, IExtOptions } from 'g-request';
import gReqeust from './gReqeust';

// IRequestOptions 类型有小程序的和 WEB 的，常用的几个属性如下，具体的话可见类型提示：

// {
//   url: string;
//   method?: IMethod;
//   header?: Record<string, string>;
//   data?: IAnyObject | ...; // 这个有多种值，xhr 这里只做做了 IAnyObject 的特殊处理，其余全部透传
//   timeout?: number; // 注意该 timeout 为默认支持的，如果确认你要兼容的都支持，那么就可以考虑去掉 ext 中的 timeout
//   ext?: Partial<IExtOptions>;
// }

// request 请求
gReqeust.request({
  url: 'xxx',
  method: 'POST', // 如不设置，默认为 GET 请求
  // 设置 header
  header: {
    Cookie: 'xxx', // 可手动设置 Cookie
  },
  data: {
    a: 1,
    b: 2,
  },
  // 可覆盖 gReqeust 实例化的 ext 设置
  ext: {
    repeatNum: 0, // 失败不重试
    taskName: 'hello', // 该请求任务名称，用于取消请求，如不设置，则默认为 gReqeust.taskIndex 的自增值
  },
});

// get 请求
gReqeust.get({
  url: 'xxx',
  // 自动拼成 query
  data: {
    a: 1,
    b: 2,
  },
  ext: {
    retcodeKey: 'code', // 这条请求的 retcodeKey 是 code 字段
  },
});

// post 请求
gReqeust.post({
  url: 'xxx',
  data: {
    a: 1,
    b: 2,
  },
  ext: {
    retcodeWhiteList: [3455, 6784], // retcode 为 0， 3455，6784 将会按成功处理，其余按失败处理
  },
});
```

**取消正在发送中的请求**

```js
// 取消单个 taskName 为 hello 的请求
gReqeust.task?.hello.abort();

// 取消所有正在发送中的请求
gReqeust.abort();
```

## ctx 参数说明

ctx 由 `req、res、ext` 三大属性组成， 其 TS 类型如下：

PS：注意所有的用于辅助功能的参数都挂到 ext 上，不要随便在 req 上面挂属性。req 会透传到请求适配器，用于发送请求的所有参数。

```ts
export interface ICtx {
  req: IReqOptions & { header: Record<string, string> }; // 请求参数
  res: IRequestSuccessCallbackResult | Record<string, never>; // 返回数据
  ext: IExtOptions; // 用于辅助的一些参数
}

export type IAdapter = (
  config: IAnyObject & { url: string },
  resolve: (value: unknown) => void,
  reject: (reason?: any) => void,
) => any;

export interface IExtOptions {
  baseUrl?: string; // 基础 url，以 https 开头
  repeatNum?: number; // 请求失败重试次数
  xRequestId?: boolean; // 是否生成请求 id
  xRequestTime?: boolean; // 是否需要记录请求时间
  timeout?: number; // 超时时间，如果确定发请求的 api 本身就支持 timeout 属性，可以设置该值为 0
  retcodeKey?: false | string; // retcode 字段，false 表示不启用该功能
  retcodeWhiteList?: false | number[]; // retcode 白名单，默认 0 和 白名单表示业务成功，其余为失败，false 表示不启用该功能。
  logicErrorMsgKey?: string; // 业务逻辑错误文本字段
  LoginErrorMsgUnknown?: string; // 默认的业务逻辑错误文本，如果后台没有返回对应的错误信息，则将使用该信息
  adapter?: IAdapter; // 请求 adapter
  [key: string]: any;
}
```

## 请求错误说明

```js
import { REQUEST_ERROR_MAP, RequestError, DEFAULT_LOGIC_ERROR_MSG_UNKNOWN } from 'g-request';

console.log(REQUEST_ERROR_MAP);

// 打印得到的错误类型有以下三种
// {
//   fail: 'REQUEST_ERROR_FAIL', // 直接 fail 的错误
//   server: 'REQUEST_ERROR_SERVER', // 服务端错误，statusCode 小于 200，大于等于 300
//   logic: 'REQUEST_ERROR_LOGIC', // 业务逻辑错误
// }
```

1、第一种错误：根本没有收到服务端返回的信息，这就有了 fail 的错误（以小程序来说，就是 fail 的回调，以 web 来说，就是 XMLHttpRequest 的 onabort, ontimeout, onerror 事件触发），该错误抛出：`new RequestError(err.message || err.errMsg, { type: REQUEST_ERROR_MAP.fail })`。

2、第二种错误：接受到了服务端的信息，但是 statusCode 小于 200，大于等于 300，这就有了 server 错误，该错误抛出： `new RequestError('Request Server Error', { type: REQUEST_ERROR_MAP.server, statusCode });`。

3、第三种错误：虽然 statusCode 大于等于 200，小于 300，但是如果用户没有登录，或身份不对等都无法获取到正确的数据，所以就有了 logic 错误，该错误抛出： `new RequestError(logicErrMsg, { type: REQUEST_ERROR_MAP.logic, retcode });`。`logicErrMsg` 的取值见下面的说明。

这三种抛出的错误默认都会进入到 `catchHandler` 进行处理，如果对 `retcodeWhiteList` 进行设置，则第三种的 logic 错误白名单内的会当做成功进入到 `thenHandler` 处理。

`catchHandler` 和 `completeHandler` 的 err 参数，就是上面的三种错误实例。

### `logicErrMsg`

`logicErrMsg` 是通过调用 `getLogicErrMsg` 方法得到的，逻辑如下：

```ts
getLogicErrMsg(ctx: ICtx): string {
  const { res, ext } = ctx;
  const { logicErrorMsgKey, loginErrorMsgUnknown } = ext;

  const defaultMsgUnknown = loginErrorMsgUnknown || DEFAULT_LOGIC_ERROR_MSG_UNKNOWN;

  // 如果没有指定 msg 的 key
  if (!logicErrorMsgKey) {
    return defaultMsgUnknown;
  }

  // 如果有指定，则取该值
  let logicErrMsg: string | undefined = res.data?.[logicErrorMsgKey];
  // 错误信息可能不是一个直接的 string 字段，而是被包裹在一个对象中，如 errData: { text: 'xxx', code: xxx };
  // 这样可以设置 key 值为 'errData.text'，分割得到最终的字段
  if (!logicErrMsg && logicErrorMsgKey.includes('.')) {
    const [key1, key2] = logicErrorMsgKey.split('.');
    if (res.data?.[key1]?.[key2]) {
      logicErrMsg = res.data[key1][key2];
    }
  }

  return logicErrMsg || defaultMsgUnknown;
}
```

## completeHandler 说明

```ts
import { Request, ICtx, RequestError, REQUEST_ERROR_MAP } from 'g-request';

// 不管成功，失败都会执行
// 这里总共有四种情况，一种成功，三种失败
gReqeust.completeHandler = (ctx: ICtx, err?: RequestError) => {
  // 成功
  if (!err) {
  }

  // 可以根据 err.type 来判断错误类型
  // ------------------------------------
  // fail
  if (err.type === REQUEST_ERROR_MAP.fail) {
  }

  // server
  if (err.type === REQUEST_ERROR_MAP.server) {
  }

  // logic
  if (err.type === REQUEST_ERROR_MAP.server) {
  }
};
```

## 其他定制

### web 请求 loading 提示

借助 [x-global-api](https://github.com/marvin1023/x-global-api) 库提供的全局 showLoading 方法实现。

如为小程序，可直接使用 `wx.showLoading` 与 `wx.hideLoading` 来实现。

```ts
import { Request, ICtx, RequestError } from 'g-request';
import { showLoading, hideLoading } from 'x-global-api';

const gReqeust = new Request({
  ext: {
    ...
    loadingTips: true,
  },
});

gReqeust.req.use((ctx) => {
  const { ext } = ctx;
  // 默认如果 loadingTips 为 true，则显示 loading
  if (ext.loadingTips) {
    showLoading({title: '数据加载中...', mask: true});
  }

  return ctx;
})

// 不管成功，失败都要 hideLoading
gReqeust.completeHandler = (ctx: ICtx, err?: RequestError) => {
  const { ext } = ctx;
  if (ext.loadingTips) {
    hideLoading();
  }
}

```

### 默认驼峰处理

以对返回的后台数据进行默认的转驼峰处理为例，在实例化的时候 `ext` 属性中添加 `camelcase`

```ts
import { Request, ICtx } from 'g-request';
import humps from 'humps';

const gReqeust = new Request({
  ext: {
    ...
    camelcase: true,
  },
});

gReqeust.res.use((ctx: ICtx) => {
  const { ext } = ctx;
  // 默认如果 camelcase 为 true，则自动进行转驼峰处理
  if (ext.camelcase) {
    ctx.res.data = humps.camelizeKeys(ctx.res.data);
  }

  return ctx;
})
```

### 小程序请求失败默认提示

```ts
import { Request, ICtx, RequestError } from 'g-request';
const gReqeust = new Request({
  ext: {
    ...
    failToast: true,
  },
});

gReqeust.catchHandler = (err: RequestError, ctx: ICtx) => {
  if (ctx.ext.failToast) {
    const { message, retcode, type, statusCode } = err || {};
    const codeText = retcode ? `(${retcode})` : '';

    if (message) {
      wx.showToast({
        title: type === 'REQUEST_ERROR_FAIL' ? '请求失败' : `${message}${codeText}`,
        icon: 'none',
      });
    }
  }

  return Promise.reject(err);
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 2.0.0 (2022-10-24)


### Features

* 导出更多函数, 方便扩展 Request 类 ([6376835](https://github.com/marvin1023/g-request/commit/63768359357604ea2dc7fd9b1e5e5fbf12f9dad2))
* 默认配置参数通过函数获取, 方便于扩展类 ([9a1d618](https://github.com/marvin1023/g-request/commit/9a1d618bc8e3e4e75f583936c23438c4cac54a1f))
* 重新梳理逻辑及 TS ([af525b2](https://github.com/marvin1023/g-request/commit/af525b2d58f2fab0a0ca622c0a6a60cdf618cbfb))
* ext 新增 urlHasNoSearch 属性,表示不带query的url ([d3a5a97](https://github.com/marvin1023/g-request/commit/d3a5a97dee2bb61d1d033caa8dd44b3d53d2fcc3))
* init ([563a1a2](https://github.com/marvin1023/g-request/commit/563a1a24e2d770af5510bb20d63d89a41f3d5a54))


### Bug Fixes

* 调整 script 及删除垃圾信息 ([8d2806a](https://github.com/marvin1023/g-request/commit/8d2806a10e61349ed143d766c7127c63778574c4))
* 默认 taskName 使用自动 id ([2a42482](https://github.com/marvin1023/g-request/commit/2a42482068838bd88a4848f44c39e398a0702f9b))
* 修复 main 入口等 ([da9c314](https://github.com/marvin1023/g-request/commit/da9c314d38342ecfd4b612863c88c4a60c121c20))


src/request/request.ts
import { IExtOptions, ICtx, IInstanceOptions, IAnyObject, IRequestOptions, IMethod, IPluginFn } from '../types';
import { REQUEST_ERROR_MAP, DEFAULT_LOGIC_ERROR_MSG_UNKNOWN, DEFAULT_RETCODE_KEY } from './const';
import { RequestError } from './requestError';
import { defaultConfig } from './default';
import { Plugins } from './plugins';
export class Request {
  options!: IInstanceOptions;
  res!: Plugins<IPluginFn, ICtx>;
  req!: Plugins<IPluginFn, ICtx>;
  task: IAnyObject = {}; // RequestTask
  taskIndex = 0; // 默认

  constructor(options?: IInstanceOptions) {
    this.init(options);
  }

  init(options?: IInstanceOptions) {
    // 非 wx.request 请求的参数全部挂在 ext 属性中
    const { ext: defaultExt } = this.getDefaultConfig();

    this.options = { ext: { ...defaultExt, ...options?.ext } };
    this.req = new Plugins<IPluginFn, ICtx>();
    this.res = new Plugins<IPluginFn, ICtx>();
  }

  // 可用于扩展类自定义新的 defaultConfig
  getDefaultConfig() {
    return defaultConfig;
  }

  // 可用于实例统一更新 ext
  // 如环境发生改变，统一更新 baseUrl
  updateOptions(options: IExtOptions) {
    Object.assign(this.options.ext, options);
  }

  initCtx(opts: IRequestOptions): ICtx {
    const { ext, ...req } = opts;

    // 构造 ctx 对象
    return {
      req: { header: {}, ...req },
      res: {},
      ext: { taskName: String(this.taskIndex), ...this.options.ext, ...ext },
    };
  }

  urlHandler(ctx: ICtx) {
    // url 处理
    const { baseUrl } = ctx.ext;
    if (baseUrl && !ctx.req.url.startsWith('http')) {
      ctx.req.url = `${baseUrl}${ctx.req.url}`;
    }
    ctx.ext.urlHasNoSearch = ctx.req.url.split('?')[0]; // 不带 query 的 url，可用于统计或上报等
  }

  reqHandler(ctx: ICtx) {
    // 自定义 id，方便打通全链路日志
    if (ctx.ext.xRequestId) {
      ctx.req.header['X-Request-Id'] = this.getUUID();
    }

    // 自定义请求发起时间
    if (ctx.ext.xRequestTime) {
      ctx.req.header['X-Request-Time'] = Date.now();
    }

    // req 插件处理入参
    this.req.pipe(ctx);
  }

  request<T>(data: IRequestOptions): Promise<T> {
    this.taskIndex++;

    const ctx = this.initCtx(data);
    this.urlHandler(ctx);
    this.reqHandler(ctx);

    return this.dispatch(ctx)
      .then(() => {
        return this.thenHandler<T>(ctx);
      })
      .catch((err: RequestError) => {
        return this.catchHandler(err, ctx);
      });
  }

  method<T>(method: IMethod, data: IRequestOptions) {
    return this.request<T>({ ...data, method });
  }

  get<T>(data: IRequestOptions) {
    return this.method<T>('GET', data);
  }

  post<T>(data: IRequestOptions) {
    return this.method<T>('POST', data);
  }

  dispatch(ctx: ICtx): Promise<IAnyObject> {
    const { timeout, taskName } = ctx.ext;

    // 重试逻辑处理
    const repeatTry = (): Promise<IAnyObject> => {
      // 超时 abort 处理
      if (timeout) {
        ctx.ext.timer = setTimeout(() => {
          this.task[taskName]?.abort();
          clearTimeout(ctx.ext.timer);
        }, timeout);
      }
      return this.promiseAdapter(ctx)
        .then((result) => {
          const { statusCode } = result;
          if (statusCode >= 200 && statusCode < 300) {
            this.clearHandler(ctx);
            // 挂到 ctx.res 上
            ctx.res = result;

            // retcode 处理
            this.retcodeHandler(ctx);

            // res 插件处理返回
            this.res.pipe(ctx);

            return this.retcodeWhiteListHandler(ctx);
          }

          // 处理服务器错误 message
          throw new RequestError('Request Server Error', { type: REQUEST_ERROR_MAP.server, statusCode });
        })
        .catch((err) => {
          // 非逻辑错误，执行清空操作及重试
          if (!(err.type && err.type === REQUEST_ERROR_MAP.logic)) {
            this.clearHandler(ctx);

            // 重试
            if (ctx.ext.repeatNum && ctx.ext.repeatNum--) {
              return repeatTry();
            }
          }

          const newError = err.type
            ? err
            : new RequestError(err.message || err.errMsg, { type: REQUEST_ERROR_MAP.fail });
          this.completeHandler(ctx, newError);
          throw newError;
        });
    };

    ctx.ext.repeatTry = repeatTry;

    return repeatTry();
  }

  promiseAdapter(ctx: ICtx): Promise<any> {
    const { req, ext } = ctx;
    const { adapter, taskName } = ext;

    if (!adapter) {
      return Promise.reject('adapter is must be required');
    }

    return new Promise((resolve, reject) => {
      this.task[taskName] = adapter(req, resolve, reject);
    });
  }

  abort() {
    Object.keys(this.task).forEach((item) => {
      this.task[item].abort();
    });
  }

  thenHandler<T>(ctx: ICtx): T {
    return ctx.res as T;
  }

  catchHandler(err: RequestError, ctx: ICtx) {
    return Promise.reject(err);
  }

  clearHandler(ctx: ICtx) {
    const { xRequestTime, timeout, timer, taskName } = ctx.ext;

    // 计算请求耗时
    if (xRequestTime) {
      ctx.ext.requestCostTime = Date.now() - ctx.req.header['X-Request-Time'];
    }

    // 删除任务
    if (this.task[taskName]) {
      delete this.task[taskName];
    }

    // 清除计时
    if (timeout && timer) {
      clearTimeout(timer);
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  completeHandler(ctx: ICtx, err?: RequestError): void | ICtx {}

  retcodeHandler(ctx: ICtx) {
    const { retcodeKey } = ctx.ext;
    // 如果key值非retcode，则手动添加 retcode 等于该 key 的值
    if (retcodeKey && retcodeKey !== DEFAULT_RETCODE_KEY) {
      ctx.res.data[DEFAULT_RETCODE_KEY] = ctx.res.data[retcodeKey];
    }

    return ctx;
  }

  retcodeWhiteListHandler(ctx: ICtx) {
    const {
      ext: { retcodeKey, retcodeWhiteList },
      res,
    } = ctx;

    // 关闭白名单，retcode 不论为啥都为成功
    // 或者没有 retcodeKey
    if (!retcodeWhiteList || !retcodeKey) {
      this.completeHandler(ctx);
      return ctx;
    }

    // 如果为 0 或者在白名单内，则进入 then 处理
    const retArr = retcodeWhiteList;
    const { retcode } = res.data || {};
    const isWhite = retArr.includes(retcode);
    if (retcode === 0 || isWhite) {
      this.completeHandler(ctx);
      return ctx;
    }

    // 逻辑错误
    const logicErrMsg = this.getLogicErrMsg(ctx);
    throw new RequestError(logicErrMsg, {
      type: REQUEST_ERROR_MAP.logic,
      retcode,
    });
  }

  getLogicErrMsg(ctx: ICtx): string {
    const { res, ext } = ctx;
    const { logicErrorMsgKey, loginErrorMsgUnknown } = ext;

    const defaultMsgUnknown = loginErrorMsgUnknown || DEFAULT_LOGIC_ERROR_MSG_UNKNOWN;

    // 如果没有指定 msg 的 key
    if (!logicErrorMsgKey) {
      return defaultMsgUnknown;
    }

    // 如果有指定，则取该值
    let logicErrMsg: string | undefined = res.data?.[logicErrorMsgKey];
    // 错误信息可能不是一个直接的 string 字段，而是被包裹在一个对象中，如 errData: { text: 'xxx', code: xxx };
    // 这样可以设置 key 值为 'errData.text'，分割得到最终的字段
    if (!logicErrMsg && logicErrorMsgKey.includes('.')) {
      const [key1, key2] = logicErrorMsgKey.split('.');
      if (res.data?.[key1]?.[key2]) {
        logicErrMsg = res.data[key1][key2];
      }
    }

    return logicErrMsg || defaultMsgUnknown;
  }

  getUUID(bytes = 16) {
    const SHARED_CHAR_CODES_ARRAY = Array(32);
    for (let i = 0; i < bytes * 2; i++) {
      SHARED_CHAR_CODES_ARRAY[i] = Math.floor(Math.random() * 16) + 48;
      // valid hex characters in the range 48-57 and 97-102
      if (SHARED_CHAR_CODES_ARRAY[i] >= 58) {
        SHARED_CHAR_CODES_ARRAY[i] += 39;
      }
    }
    return String.fromCharCode.apply(null, SHARED_CHAR_CODES_ARRAY.slice(0, bytes * 2));
  }
}
